### DISCUSSION TITLE
D-bus path extracted from web page

### DISCUSSION THREAD

**Discussion by seamaster101** [Louisvdw/dbus-serialbattery#626] [General]
I have my batteries' BMSs talking to my own ESP32 and displaying webpage that looks like this:
<img width="567" alt="image" src="https://user-images.githubusercontent.com/7897023/236733453-7ab85315-f622-473e-964a-b4d85e9807bc.png">

The ESP32 is always connected to the WiFi so potentially script running on the RPI / Venus OS, can parse the values and display them .  
The question is how easy is to use POST or GET command to create new D-Bus path for my batteries. (unfortunately they don't have one of the known BMSs that is supported by dbus-serialbattery :( )
Is anyone have written and script to pull battery information from web page?
The protocol for my batteries is described here and if it is easier to implement that rather than parsing the webpage, please give me some hints.
I have posted the actual serial protocol here https://github.com/Louisvdw/dbus-serialbattery/discussions/486 before, but unfortunately no one answered.

**Comment by mr-manuel** [comment]
Hi,

you can feed the [venus-os_dbus-mqtt-battery](https://github.com/mr-manuel/venus-os_dbus-mqtt-battery) over MQTT. Probably with Node-RED you can fetch the values from your webpage and send them over MQTT. There are a lot of how to's for how to do it.

Node-RED is included in the Venus OS large.

For your request https://github.com/Louisvdw/dbus-serialbattery/discussions/486 there is not enough common interrest actually.

**Comment by seamaster101** [comment]
@mr-manuel thank you for the reply. As far as #486 is concerned I totally understand that. 
I was hoping that you can give me some more directions for  `https://github.com/mr-manuel/venus-os_dbus-mqtt-battery`. 
I actually have a flow that parses my webpage in node-red and sends it to influxDB. Starting from this script, I spent may hours last night trying to make it publish to MQTT. I have to admit, I'm not very familiar with the whole MQTT thingy.... 
[Parse Batteries.zip](https://github.com/Louisvdw/dbus-serialbattery/files/11426580/Parse.Batteries.zip) 
Attached is the flow that parses my batteries and gives all of the information. What will take you 5 min to figure out will for sure take me days of frustrations...
Can you please help?
<img width="1103" alt="image" src="https://user-images.githubusercontent.com/7897023/236978268-e1c55985-647d-4a2a-ba6d-1b9aab23beba.png">

**Reply by mr-manuel** [reply]
See https://github.com/mr-manuel/venus-os_dbus-mqtt-battery/discussions/10#discussioncomment-5693581.

You only need to create a JS Object, like you did in the function node with [this structure](https://github.com/mr-manuel/venus-os_dbus-mqtt-battery#json-structure) and send it to a MQTT out node. The broker and topic from the out node has to be the same as in the config of the MQTT Battery config.

As example, attach an additional function node (one per battery) after the existing function node with this code (insert it in the new function node):

**Battery 1**
```javascript
msg.payload = {
    "Dc": {
        "Power": msg.payload["data"][0]["fields"]["Power"],
        "Voltage": msg.payload["data"][0]["fields"]["Voltage"],
        "Current": msg.payload["data"][0]["fields"]["Current"],
        "Temperature": msg.payload["data"][0]["fields"]["Temperature"],
    },
    "Soc": msg.payload["data"][0]["fields"]["SOC"],
}
```

**Battery 2**
```javascript
msg.payload = {
    "Dc": {
        "Power": msg.payload["data"][1]["fields"]["Power"],
        "Voltage": msg.payload["data"][1]["fields"]["Voltage"],
        "Current": msg.payload["data"][1]["fields"]["Current"],
        "Temperature": msg.payload["data"][1]["fields"]["Temperature"],
    },
    "Soc": msg.payload["data"][1]["fields"]["SOC"],
}
```

This is only an example and there are a lot more fields you can fill. All is found in the JSON structure.

As this discussion is for dbus-serialbattery and not dbus-mqtt-battery would it be better, to open a new discussion in the right repository.
